Bootstrap cross-capi-metal demo

hasheddan opened the pull request:
Initial setup with placeholders and demo steps for discussion.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan commented on README.md:
```diff
@@ -1,2 +1,43 @@
+
+1. Create a control plane cluster. This can be any Kubernetes cluster. We
+   recommend starting out with [KIND](https://kind.sigs.k8s.io/).
+2. Install CAPI's
+   [clusterctl](https://cluster-api.sigs.k8s.io/user/quick-start.html#install-clusterctl).
+3. Go through the [CAPI
+   quickstart](https://cluster-api.sigs.k8s.io/user/quick-start.html) to
+   provision a cluster on Equinix Metal (still called `Packet` in CAPI docs).
```
@detiber ideally we skip the `clusterctl` quickstart and instead put the types that get created into a composition so we can do everything in what is currently step 8. Obviously there would be some setup still required to have CAPI and the equinix metal provider present, but once that is true do you know what CRs are created (i.e. what we could move to the composition)?

detiber replied:
There should be a way that we can dump out all the yaml that is applied by clusterctl, will work up the details.

detiber replied:
Dumping provider manifests:
```sh
clusterctl config provider --core=cluster-api -o yaml > core.yaml
clusterctl config provider --bootstrap=kubeadmi -o yaml > kubeadm-bootstrap.yaml
clusterctl config provider --control-plane=kubeadmi -o yaml > kubeadm-control-plane.yaml
clusterctl config provider --infrastructure=packet -o yaml > packet-infrastructure.yaml
```

The output above will need to be run through envsubst before being applied, the main env var that we'd need to worry about is PACKET_API_KEY, we should be able to rely on the defaults for the rest.

Prior to applying the above yaml, cert-manager will need to be deployed on the management cluster.

`clusterctl config cluster` can be used similar to the quickstart to pre-generate the yaml needed for creating a cluster, unlike `clusterctl init` it doesn't operate directly against a k8s cluster it just outputs the processed template to stdout.

Once the cluster yaml is applied, then you should be able to look for the presence of the kubeconfig secret, which will live in the same namespace as the cluster created and be named `<cluster name>-kubeconfig`.

Once the kubeconfig is retrieved, it can be used to poll the remote cluster for availability, once it can be reached, then the CNI can be deployed (cluster configuration defaults are aligned with calico).

Once CNI is deployed, any additional control plane instances and worker node instances will be able to be created.

hasheddan replied:
I think https://cluster-api.sigs.k8s.io/user/quick-start.html#generating-the-cluster-configuration is the main step we want to do in composition, the rest can just be setup beforehand

hasheddan commented on charts/echo/Chart.yaml:
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+appVersion: 0.0.1
+description: Equinix echo is an example echo client / server application.
+name: equinix-echo
```
Definitely open to a better demo use case that shows connectivity between cluster and machine if y'all have ideas.

detiber replied:
We could probably hack something from this list to work: https://www.virtuallyghetto.com/2020/06/interesting-kubernetes-application-demos.html

There is a pacman app that was created for use with federation demos that we could also likely hack up pretty easily for our use: https://github.com/font/k8s-example-apps

hasheddan replied:
Looks like there are some interesting ones there. From what I can tell they mostly involve both places running kubernetes, and I think in this demo only the CAPI side will be right?

detiber replied:
Yeah, I think any would require some modification for sure, but anything that has a database component is something that we could move out of k8s relatively easily for our purposes, I think.

hasheddan replied:
@detiber so are you thinking run database component on crossplane provisioned machine and the app on the CAPI cluster?

detiber replied:
@hasheddan yeah, that is what I was thinking

hasheddan commented on package/machine/composition.yaml:
```diff
@@ -0,0 +1 @@
+# Composition with Equinix Metal Machine / Virtual Network / Assignment
```
@displague will this be the items we need to get an IP address for the machine? If the cluster and the machine are in the same virtual network I assume we won't need a public IP for the machine?

displague replied:
The XP provisioned device will have a 10.x.x.x private address, and IPv6 public address, and an IPv4 public address by default.

We have additional options to provision devices with additional dedicated addresses or EIP addresses (or ranges):
* https://github.com/packethost/crossplane-provider-equinix-metal/blob/master/apis/server/v1alpha2/device_types.go#L176
* https://github.com/packethost/crossplane-provider-equinix-metal/blob/master/apis/server/v1alpha2/device_types.go#L141
* https://github.com/packethost/crossplane-provider-equinix-metal/blob/master/apis/server/v1alpha2/device_types.go#L163

We also have the option of adding unmanaged addresses to each node which will be available over a vlan.

---

I'm not sure which of these features are available in the cluster-api-provider.

hasheddan replied:
@displague thanks for the info! Seems like simplest solution would be to just have Machine be publicly reachable and client in the CAPI cluster pinging the server on the Machine. However, would be pretty cool for Crossplane to create the virtual network and for CAPI to deploy the cluster in it. @detiber do you know if there is support for this?

detiber replied:
Currently we only support the default (public) networking with cluster-api-provider-packet.

hasheddan replied:
Sounds like we will just do over public internet then.

hasheddan committed:
Build crossplane package for CAPI / Equinix metal demo

The Configuration package creates a CAPI cluster on Equinix Metal,
deploys a separate Device with MySQL installed, and deploys a Helm chart
into the CAPI cluster.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Simple application for interacting with MySQL DB

Adds a simple application that allows users to read and write to a MySQL
database.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Bootstrap Helm chart for application

Adds helm chart for deploying the demo application. Currently missing
secret required to talk to MySQL DB.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Update README.md with proposed workflow

Updates steps in README.md to include expanded functionality of the
cross-capi-metal package.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>
